DirectProgramming/DPC++FPGA/Tutorials/Features/fpga_reg/src/fpga_reg.cpp

```diff
@@ -0,0 +1,216 @@
+          #pragma unroll
+          for (size_t j = 0; j < kSize; j++) {
+#ifdef USE_FPGA_REG
+            // Use fpga_reg to insert a register between the copy of val used
+            // in each unrolled iteration. 
+            val = intel::fpga_reg(val);
+            // Since val is held constant across the kSize unrolled iterations,
+            // the FPGA hardware structure of val's distribution changes from a
+            // kSize-way fanout (without fpga_reg) to a chain of of registers  
+            // with intermediate tap offs. Refer to the diagram in the README.
+  
+            // Use fpga_reg to insert a register between each step in the acc
+            // adder chain.
+            acc = intel::fpga_reg(acc) + (coeff[j] * (val + kOffset[j]));
+            // This transforms a compiler-inferred adder tree into an adder 
+            // chain, altering the structure of the pipeline. Refer to the 
+            // diagram in the README.
+#else
+            // Without fpga_reg, the compiler schedules the operations here
+            // according to its default optimization heuristics.
+            acc += (coeff[j] * (val + kOffset[j]));
+#endif
+          }
+
+          // Rotate the values of the coefficient array.
+          // The loop is fully unrolled. This is a cannonical code structure;
+          // the DPC++ compiler's FPGA backend infers a shift register here.          
+          int tmp = coeff[0];
+          #pragma unroll
+          for (size_t j = 0; j < kSize - 1; j++) {
+            coeff[j] = coeff[j + 1];
+          }
+          coeff[kSize - 1] = tmp;
+
+          // Result
+          r[i] = acc;
+        }
+      });
+    });
+
+    // Measure kernel execution time
+    double start = e.get_profiling_info<info::event_profiling::command_start>();
+    double end = e.get_profiling_info<info::event_profiling::command_end>();
+    // Convert from nanoseconds to milliseconds.
+    double kernel_time = (end - start) * 1e-6;
+
+    // Kernel consists of two nested loops with 3 operations in the innermost
+    // loop: 2 additions and 1 multiplication operation.
+    size_t num_ops_per_kernel = input_size * kSize * 3;
+    cout << "Throughput for kernel with input size " << input_size
+         << " and coefficient array size " << kSize << ": ";
+    cout << std::fixed << std::setprecision(6)
+         << ((double)num_ops_per_kernel / kernel_time) / 1.0e6 << " GFlops\n";
+         
+  } catch (sycl::exception const &e) {
+    // Catches exceptions in the host code
+    std::cout << "Caught a SYCL host exception:\n" << e.what() << "\n";
+
+    // Most likely the runtime couldn't find FPGA hardware!
+    if (e.get_cl_code() == CL_DEVICE_NOT_FOUND) {
+      std::cout << "If you are targeting an FPGA, please ensure that your "
+                   "system has a correctly configured FPGA board.\n";
+      std::cout << "If you are targeting the FPGA emulator, compile with "
+                   "-DFPGA_EMULATOR.\n";
```

Comment:
You might never see the message if you never use `std::endl` or `std::flush`

Reply:
So, what I mean is to replace `\n` here by `std::endl`

Reply:
Yes, I agree explicit flush is needed here. The use of '\n' over std::endl is another requirement of the coding standard laid out for Intel-authored samples. I need to initiate a discussion to have the recommendations amended.